File: mask-python/mask_privacy/core/span.py
"""
Span Resolution Engine — Sweep-Line Overlap Resolver.

Instead of mutating the text string after every regex hit (O(N·hits) allocations),
all tiers now collect lightweight ``Span`` objects.  Once all tiers have finished,
``resolve_overlaps`` performs a single O(K log K) pass to choose the winning span
in every conflicting region, and the caller rebuilds the string exactly once.

Algorithm
---------
1. Sort spans by (start ASC, length DESC, confidence DESC).
2. Walk left-to-right, tracking the furthest consumed index ``occupied_end``.
3. Any span that starts before ``occupied_end`` is a conflict:
   - Fully contained → always discard (champion already covers it).
   - Partial overlap  → kept only if its confidence is strictly higher; the
     champion's end is used as the trim boundary (we don't split tokens).
4. The surviving spans are returned in reverse start order so the caller can
   replace right-to-left without offset drift.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import List, Optional


@dataclass(order=False)
class Span:
    """Represents one PII match from any detection tier."""

    start: int
    end: int
    entity_type: str
    original_value: str
    confidence: float
    method: str                          # "dlp_heuristic" | "regex" | "nlp"
    language: Optional[str] = field(default=None)

    # Filled in by the caller after encode_fn is applied
    masked_value: Optional[str] = field(default=None)

    @property
    def length(self) -> int:
        return self.end - self.start


def resolve_overlaps(spans: List[Span]) -> List[Span]:
    """Return a non-overlapping, right-to-left-ordered subset of *spans*.

    Parameters
    ----------
    spans:
        Raw, potentially overlapping list from all scanning tiers.

    Returns
    -------
    List[Span]
        De-duplicated spans sorted by ``start`` descending — ready for
        right-to-left string reconstruction.
    """
    if not spans:
        return []

    # Primary sort: start ASC; secondary: length DESC (prefer longer spans);
    # tertiary: confidence DESC (prefer more certain matches).
    sorted_spans = sorted(
        spans,
        key=lambda s: (s.start, -(s.end - s.start), -s.confidence),
    )

    resolved: List[Span] = []
    occupied_end: int = -1

    for span in sorted_spans:
        if span.start >= occupied_end:
            # No overlap — accept this span unconditionally.
            resolved.append(span)
            occupied_end = span.end
        elif span.end <= occupied_end:
            # Fully inside an already-accepted span — discard.
            continue
        else:
            # Partial overlap: span starts inside occupied zone but extends
            # beyond it.  Only keep if it is more confident than the current
            # champion; we purposely do NOT split the overlapping tokens.
            last = resolved[-1]
            if span.confidence > last.confidence:
                resolved.pop()
                resolved.append(span)
                occupied_end = span.end
            # else: discard the challenger

    # Return in descending start order so callers can slice right-to-left
    # without offset drift.
    resolved.sort(key=lambda s: s.start, reverse=True)
    return resolved


def reconstruct(text: str, resolved_spans: List[Span]) -> str:
    """Rebuild *text* from a right-to-left-ordered list of resolved spans.

    Each span must have ``masked_value`` already populated.

    This is the single string-construction pass that replaces all 50+ per-hit
    allocations in the old waterfall pipeline.
    """
    # resolved_spans must be sorted start DESC already (output of resolve_overlaps)
    result = text
    for span in resolved_spans:
        if span.masked_value is None:
            continue
        result = result[: span.start] + span.masked_value + result[span.end :]
    return result

File: mask-ts/src/core/span.ts
/**
 * Span Resolution Engine — Sweep-Line Overlap Resolver (TypeScript).
 *
 * All detection tiers now return Span objects instead of mutating the text.
 * resolveOverlaps() chooses the winning span in every conflicting region,
 * and reconstruct() rebuilds the string exactly once.
 */

export interface Span {
  start: number;
  end: number;
  entityType: string;
  originalValue: string;
  confidence: number;
  method: string;      // "dlp_heuristic" | "regex" | "nlp"
  language?: string;
  // Audit trail fields (DSPM compliance)
  ruleId?: string;                 // e.g. "MASK-FIN-001"
  complianceScope?: ReadonlySet<string>;  // e.g. {"PCI-DSS", "HIPAA"}
  maskedValue?: string;
}

/**
 * Return a non-overlapping, right-to-left-ordered subset of spans.
 *
 * Algorithm:
 * 1. Sort by start ASC, length DESC (prefer longer), confidence DESC.
 * 2. Walk left-to-right tracking occupiedEnd.
 * 3. Fully-contained spans are discarded.
 * 4. Partial overlaps resolve by confidence (higher wins).
 */
export function resolveOverlaps(spans: Span[]): Span[] {
  if (spans.length === 0) return [];

  const sorted = [...spans].sort((a, b) => {
    if (a.start !== b.start) return a.start - b.start;
    const lenDiff = (b.end - b.start) - (a.end - a.start);
    if (lenDiff !== 0) return lenDiff;
    return b.confidence - a.confidence;
  });

  const resolved: Span[] = [];
  let occupiedEnd = -1;

  for (const span of sorted) {
    if (span.start >= occupiedEnd) {
      resolved.push(span);
      occupiedEnd = span.end;
    } else if (span.end <= occupiedEnd) {
      // Fully inside an already-accepted span — discard.
      continue;
    } else {
      // Partial overlap — keep highest confidence.
      const last = resolved[resolved.length - 1];
      if (span.confidence > last.confidence) {
        resolved.pop();
        resolved.push(span);
        occupiedEnd = span.end;
      }
    }
  }

  // Return descending start order for right-to-left reconstruction.
  return resolved.sort((a, b) => b.start - a.start);
}

/**
 * Rebuild text from a right-to-left-ordered list of resolved spans.
 * Each span must have maskedValue populated.
 * This is the single string-construction pass that replaces all slice loops.
 */
export function reconstruct(text: string, resolvedSpans: Span[]): string {
  let result = text;
  for (const span of resolvedSpans) {
    if (span.maskedValue == null) continue;
    result = result.slice(0, span.start) + span.maskedValue + result.slice(span.end);
  }
  return result;
}
